Fix/tcp delay causing composite tcp packets

thosaa opened the pull request:
I was investigating a problem and noticed that we could trigger the modbus lib to 'collect' multiple Modbus ADUs into a single TCP PDU. 

This is from Wireshark showing the trace between the client (a nodejs script for testing purposes) and the server (baked with tokio-modbus): 
![image](https://user-images.githubusercontent.com/12253631/201349992-3564eb50-fb2d-451a-bd56-26fb8f9a56f9.png)

Notice that packet 113 has multiple responses, which can be viewed:
![image](https://user-images.githubusercontent.com/12253631/201350254-3feaac8b-70cb-44b8-aaa4-a3c03473b241.png)

I am pretty sure that this behavior is causing a problem for one of our customers, who is using a PLC which does support a list of Modbus ADUs in a single TCP PDU. In the other cases our customers PLCs must be waiting for the reponse before making a new request. 

From the [Modbus_Messaging_Implementation_Guide_V1_0b.pdf](http://modbus.org/docs/Modbus_Messaging_Implementation_Guide_V1_0b.pdf) in your README, it seems that this should not be done at all. 
At page 10:
![image](https://user-images.githubusercontent.com/12253631/201351736-41db465c-3c0b-4b04-ae29-9b95fd34dc6c.png)

This fix is simply to disable Nagle's algorithm by setting nodelay = true. This can not be done the tokio::net::TcpListener directly so I was forced to use from_std and constructing a standard socket.

I was considering making the nodelay a setting, but as it advised not to use it in the document above, I think it should be standard.

This is what it looks like after the fix.
Note that packet 102 / Transaction 5 has 3 requests (Modbus ADUs) in it. The Decoder is perfectly cable of handling this response .
![image](https://user-images.githubusercontent.com/12253631/201348813-d85c0f00-7779-457f-8bbf-8aa630c6abd1.png)

flosse commented on src/server/tcp.rs:
```diff
@@ -43,7 +43,12 @@ impl Server {
         let service = Arc::new(service);
-        let listener = TcpListener::bind(self.socket_addr).await?;
+        let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
+        socket.reuse_address()?;
```
What is this for?
Do you mean `socket.set_reuse_address(true)?`

flosse commented:
Would you mind to fix the CI issues and add a note to the `CHANGELOG.md`?

thosaa commented:
> Would you mind to fix the CI issues and add a note to the `CHANGELOG.md`?
Fixed fmt.
I added a note and guessed it should be a new release, let me know if it should not.

flosse commented on src/server/tcp.rs:
```diff
@@ -43,7 +43,11 @@ impl Server {
     {
         let service = Arc::new(service);
-        let listener = TcpListener::bind(self.socket_addr).await?;
+        let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
```
Why is the `protocol` parameter set to `None`? What would happen if we use `Protocol::TCP`?

flosse commented on src/server/tcp.rs:
```diff
@@ -43,7 +43,11 @@ impl Server {
+        let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
+        socket.set_nodelay(true)?;
+        socket.bind(&self.socket_addr.into())?;
+        socket.listen(1024)?;
```
Could you add some notes why `1024` is used? Ideally point to some documentation somewhere.

uklotzde replied:
A (private) constant with comments would be appropriate here.

thosaa replied:
The param to listen is the number of connections allowed queued as I read it from [listen(2)](https://man7.org/linux/man-pages/man2/listen.2.html)
I dont think it's the number of allowed simultanious connections, just "not-yet-accepted" connections. 

I have added a const with comments

flosse commented:
> > Would you mind to fix the CI issues and add a note to the `CHANGELOG.md`?

> Fixed fmt.
> I added a note and guessed it should be a new release, let me know if it should not.

Thanks!

uklotzde commented:
@flosse We should squash the commits before merging.

thosaa commented on src/server/tcp.rs:
```diff
@@ -21,6 +21,9 @@ use std::{
 use tokio::net::{TcpListener, TcpStream};
 use tokio_util::codec::Framed;
 
+/// The maximum number of connections queue for acception 
```
Doh! Of cause :-)

uklotzde commented:
This PR is obsolete now after #159 has been merged. The TCP server accepts a `TcpListener` which could be configured by the application.
